packages/types/src/types.d.ts

```diff
@@ -27,12 +28,43 @@ export type SnapRpcHandler = (args: {
 
 export type OnRpcRequestHandler = SnapRpcHandler;
 
+export type OnTransactionResponse = {
+  insights: { [key: string]: unknown };
+};
+
+// @todo improve type
+export type OnTransactionHandler = (args: {
+  origin: string;
+  transaction: { [key: string]: unknown };
+  chainId: ChainId;
+}) => Promise<OnTransactionResponse>;
+
+export type SnapRpcHookArgs = {
+  origin: string;
+  handler: HandlerType;
+  request: Record<string, unknown>;
+};
+
+// The snap is the callee
+export type SnapRpcHook = (options: SnapRpcHookArgs) => Promise<unknown>;
+
 export type SnapProvider = MetaMaskInpageProvider;
 
-export type SnapId = string;
+export type ChainId = string;
```

Comment:
All of the added types in this range should be moved to `/snap-utils` because they are internal.